modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusStateUpdater.scala

```diff
@@ -58,52 +58,67 @@ object ConsensusStateUpdater {
 
     private val logger = Slf4jLogger.getLoggerFromClass(ConsensusStateUpdater.getClass)
 
+    /** Pair of (maybeUpdatedState, effect) */
+    type ModifyStateResult = (Option[ConsensusState[Key, Artifact]], F[Unit])
+
     def tryFacilitateConsensus(
       key: Key,
       lastKeyAndArtifact: (Key, Artifact)
     ): F[Option[ConsensusState[Key, Artifact]]] =
       consensusStorage
-        .updateState(key)(logStatusIfUpdated(internalTryFacilitateConsensus(key, lastKeyAndArtifact)))
+        .condModifyState(key)(toModifyStateFn(internalTryFacilitateConsensus(key, lastKeyAndArtifact)))
+        .flatMap(evalEffect)
+        .flatTap(logIfModified(key))
 
     def tryAdvanceConsensus(
       key: Key,
       resources: ConsensusResources[Artifact]
     ): F[Option[ConsensusState[Key, Artifact]]] =
       consensusStorage
-        .updateState(key)(logStatusIfUpdated(internalTryAdvanceConsensus(key, resources)))
-
-    import consensusStorage.StateUpdateFn
-
-    private def logStatusIfUpdated(fn: StateUpdateFn): StateUpdateFn =
-      fn(_, _).flatTap {
-        case Some((true, newState)) =>
-          logger.debug(s"Consensus status for key ${newState.key.show} transitioned to ${newState.status.show}")
-        case _ => Applicative[F].unit
+        .condModifyState(key)(toModifyStateFn(internalTryAdvanceConsensus(key, resources)))
+        .flatMap(evalEffect)
+        .flatTap(logIfModified(key))
+
+    import consensusStorage.ModifyStateFn
+
+    private def toModifyStateFn(
+      fn: Option[ConsensusState[Key, Artifact]] => F[Option[(ConsensusState[Key, Artifact], F[Unit])]]
+    ): ModifyStateFn[ModifyStateResult] =
+      maybeState =>
+        fn(maybeState).map(_.map {
+          case (state, effect) => (state.some, (state.some, effect))
+        })
+
+    private def evalEffect(result: Option[ModifyStateResult]): F[Option[ConsensusState[Key, Artifact]]] =
+      result.flatTraverse {
+        case (maybeState, effect) => effect >> maybeState.pure[F]
```

Comment:
`effect.map(_ => maybeState)`

Reply:
what about error in `effect`? shouldn't it be handled here?

Reply:
So there's very little we can do at this point if an error happens, except for logging the error, and we already do it as a part of gossip error handling. This effect is evaluated after the state transition is "committed" and some effects have already taken place. The only sensible error handling would be to voluntarily leave the consensus, which isn't implemented and possibly will not have to be implemented if the node gets kicked out by the health-check anyway.

Reply:
Didn't we agree yesterday that any error in effect during state transition should switch state to its previous/initial state?

Reply:
As stated above
>This effect is evaluated after the state transition is "committed"